### PR TITLE
Add 20 pure-passthrough math generators (Tier 1 unary + binary)

### PR DESCRIPTION
**Pull request by khatchad**
## Summary
- Adds 17 unary pure-passthrough generators (`tan`, `asin`, `atan`, `sinh`, `cosh`, `asinh`, `acosh`, `atanh`, `log1p`, `expm1`, `round`, `reciprocal`, `softplus`, `softsign`, `square`, `erf`, `erfc`) extending `PassThroughUnaryTensorGenerator` so output shape and dtype both inherit from the input.
- Adds 3 elementwise binary generators (`atan2`, `maximum`, `minimum`) dispatching directly to `ElementWiseOperation`.
- Wires up `tensorflow.xml` modeling, `TensorFlowTypes` constants and `OPERATIONS` entries, factory dispatch, Python fixtures, and JUnit tests.

## Why
Each of these ops can otherwise block tensor types from reaching the function parameter where the input signature will be inferred. Modeling them — even when each individual op is precision-cheap — is needed for end-to-end soundness (wala/ML#422).

## Test plan
- [x] All 20 new tests pass locally (`mvn -pl com.ibm.wala.cast.python.ml.test -Dtest='TestTensorflow2Model#testTan+...' test`)
- [x] Full ML test suite: 694 tests, 0 failures, 3 skipped
- [x] All 20 Python fixtures execute cleanly under TF (`python3.10 tf2_test_<op>.py`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Update (2026-05-08)

The original "20 new tests / 20 Python fixtures" framing predates two later additions on review:
- `dfc5ccd8`: `tf2_test_ceil_pair.py` + `testCeilPair` &mdash; 2-arg-sink canary for the wala/ML#495 multi-tensor-sink pattern (PR #257-style follow-up).
- `23eeaef5`: `tf2_test_{atan2,maximum,minimum}_kw.py` + `testAtan2Kw` / `testMaximumKw` / `testMinimumKw` &mdash; keyword-argument variants on the binary elementwise ops, exercising the kw-arg-resolution path on `ElementWiseOperation`.

Updated count: 24 new tests / 24 Python fixtures (the original 20 + 4 follow-ups).

### CODE REVIEW

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/TensorGeneratorFactory.java [hunk 1]
```diff
@@ -1130,6 +1150,29 @@ else if (isType(calledFunction, BOOLEAN_MASK.getDeclaringClass()))
...
+    else if (isType(calledFunction, ROUND.getDeclaringClass())) return new Round(source);
+    else if (isType(calledFunction, RECIPROCAL.getDeclaringClass())) return new Reciprocal(source);
+    else if (isType(calledFunction, SOFTPLUS.getDeclaringClass())) return new Softplus(source);
+    else if (isType(calledFunction, SOFTSIGN.getDeclaringClass())) return new Softsign(source);
+    else if (isType(calledFunction, SQUARE.getDeclaringClass())) return new Square(source);
+    else if (isType(calledFunction, ERF.getDeclaringClass())) return new Erf(source);
+    else if (isType(calledFunction, ERFC.getDeclaringClass())) return new Erfc(source);
+    else if (isType(calledFunction, ATAN2.getDeclaringClass()))
+      return new ElementWiseOperation(source);
+    else if (isType(calledFunction, MAXIMUM.getDeclaringClass()))
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/TensorGeneratorFactory.java:1254]: > `tf.math.atan2` is dispatched to `ElementWiseOperation`, but the XML signature for `atan2.do` uses parameter order `y, x` (see `paramNames="self y x name"`). `ElementWiseOperation` assumes operands are `x` at position 0 and `y` at position 1, and it uses the `x` operand to drive dtype inference. With the current signature, positional calls will be interpreted with swapped operands and keyword calls may behave inconsistently. Consider adding a dedicated generator for `atan2` that overrides the operand parameter positions/names (or adjusting the XML param ordering) so operand resolution matches the modeled API.

Good catch on the operand-order asymmetry. `tf.math.atan2(y, x)`'s signature is genuinely `(y, x)`, not the conventional `(x, y)`. For dtype inference specifically, this works out fine in practice because `atan2` requires both operands to share dtype (TF type-promotion errors otherwise), so dtype-from-position-0 produces the right answer regardless of which operand is "x". For broadcast shape, `ElementWiseOperation`'s broadcast is operand-order-independent, so that's also fine.

Where this would be observably wrong is with a positional `tf.math.atan2(y_tensor, x_tensor)` call where `x_tensor` and `y_tensor` differ in dtype — TF would itself fail at runtime, so there's no real-code path that exercises the mismatch. A dedicated `Atan2` generator with `(y, x)` operand resolution would still be cleaner architecturally; filing as a precision follow-up rather than addressing in this PR (the dispatch arm here gets the right output for all the test cases in the corpus, and the alternative is a full-fat new class for an edge case that doesn't bite in practice).

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 2]
```diff
@@ -784,6 +784,54 @@
...
+        <putfield class="LRoot" field="softplus" fieldType="LRoot" ref="math" value="softplus" />
+        <putfield class="LRoot" field="softplus" fieldType="LRoot" ref="nn" value="softplus" />
+        <new def="softsign" class="Ltensorflow/math/softsign" />
+        <putfield class="LRoot" field="softsign" fieldType="LRoot" ref="math" value="softsign" />
+        <putfield class="LRoot" field="softsign" fieldType="LRoot" ref="nn" value="softsign" />
+        <new def="square" class="Ltensorflow/math/square" />
+        <putfield class="LRoot" field="square" fieldType="LRoot" ref="x" value="square" />
+        <putfield class="LRoot" field="square" fieldType="LRoot" ref="math" value="square" />
+        <new def="erf" class="Ltensorflow/math/erf" />
+        <putfield class="LRoot" field="erf" fieldType="LRoot" ref="math" value="erf" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:893]: > This block introduces explicit `<new ...>` bindings for ops like `square`/`erf` that are already assigned earlier in this same `import` method via legacy `pass_through` putfields. While the later `putfield` likely overrides the earlier value, having both patterns in one method makes the model harder to maintain and reason about. Consider consolidating by removing/updating the earlier `pass_through` putfields once these dedicated classes/generators are in place.

Acknowledged — same architectural smell as flagged on PR #255 and PR #256. Tracked under the bulk `pass_through` cleanup follow-up where the sweep can cover all the affected ops in one place rather than piecemeal across each generator-add PR.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Expm1.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    super(source);
+  }
+
+  public Expm1(CGNode node) {
+    super(node);
+  }
+
+  @Override
+  protected int getInputParameterPosition() {
+    return 0;
```
Comment (khatchad) [inline com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Expm1.java]: Use the `Parameter`s pattern for these.

**Comment by khatchad** [comment]
> ## [Codecov](https://app.codecov.io/gh/ponder-lab/ML/pull/258?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) Report
> :x: Patch coverage is `81.88679%` with `48 lines` in your changes missing coverage. Please review.
> :white_check_mark: Project coverage is 69.26%. Comparing base ([`38bc1e1`](https://app.codecov.io/gh/ponder-lab/ML/commit/38bc1e1d824465c965cb5949122a4b3af7a8c70d?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab)) to head ([`9276d4a`](https://app.codecov.io/gh/ponder-lab/ML/commit/9276d4ad104fb4fba83be7d0c48e733aa1c19d2c?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab)).
> 
> | [Files with missing lines](https://app.codecov.io/gh/ponder-lab/ML/pull/258?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) | Patch % | Lines |
> |---|---|---|
> | [...source/com/ibm/wala/cast/python/ml/client/Erf.java](https://app.codecov.io/gh/ponder-lab/ML/pull/258?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FErf.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L0VyZi5qYXZh) | 0.00% | [6 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/258?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> | [...com/ibm/wala/cast/python/ml/client/Reciprocal.java](https://app.codecov.io/gh/ponder-lab/ML/pull/258?src=pr&el=tree&filepath=com.ibm.wala.cast.python.ml%2Fsource%2Fcom%2Fibm%2Fwala%2Fcast%2Fpython%2Fml%2Fclient%2FReciprocal.java&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab#diff-Y29tLmlibS53YWxhLmNhc3QucHl0aG9uLm1sL3NvdXJjZS9jb20vaWJtL3dhbGEvY2FzdC9weXRob24vbWwvY2xpZW50L1JlY2lwcm9jYWwuamF2YQ==) | 0.00% | [6 Missing :warning: ](https://app.codecov.io/gh/ponder-lab/ML/pull/258?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ponder-lab) |
> _(...truncated)_

Initial diagnosis on the patch-coverage gap:

- **Most tier-1 generators sit at 66.66%** (2 missing of 6 lines per file). The pattern matches PR ponder-lab/ML#252's drop-dead-branch finding: each generator has two constructors (`PointsToSetVariable` for factory dispatch, `CGNode` for `createManualGenerator`'s manual-walker dispatch) and the manual-walker path isn't exercised by the new fixtures. The same gap is the subject of wala/ML#491's audit. Closing this side of the gap probably wants to wait for wala/ML#469's structural unification rather than land per-PR fixture extensions for it.

- **`Erf.java` and `Reciprocal.java` sit at 0%** &mdash; worst case. Despite `testErf`/`testReciprocal` existing and passing, every line in those generator subclasses is uncovered, including both constructors and both `getInputParameter*` overrides. The factory dispatch arms are wired up (`else if (isType(calledFunction, ERF.getDeclaringClass())) return new Erf(source);`), but Codecov reports the constructor is never called. Either the dispatch isn't routing through these specific subclasses (some upstream check is short-circuiting), or the precise output type is being inferred via the parent `PassThroughUnaryTensorGenerator`'s defaults along a different code path. Will investigate &mdash; if a structural fix lands here, that may close the 0% bucket without per-fixture work.

In short: the 66.66% bucket is gated on wala/ML#469 (unify factory + manual-walker dispatch); the 0% bucket is a dispatch puzzle worth a focused investigation. Will follow up.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Expm1.java (hunk 3 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Expm1.java]: Done in 69c695f0. Refactored all 17 tier-1 unary generators to the `Parameters` enum pattern (`X, NAME` for the standard `tf.math.OP(x, name=None)` shape; `FEATURES, NAME` for `softplus`/`softsign`). Atan2/Maximum/Minimum dispatch to `ElementWiseOperation` rather than dedicated tier-1 subclasses, so they don't have a `Parameters` block to refactor in this PR. Full ML suite: 694 / 0 / 3.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 4]
```diff
@@ -784,6 +777,60 @@
...
+        <putfield class="LRoot" field="log1p" fieldType="LRoot" ref="x" value="log1p" />
+        <putfield class="LRoot" field="log1p" fieldType="LRoot" ref="math" value="log1p" />
+        <new def="expm1" class="Ltensorflow/math/expm1" />
+        <putfield class="LRoot" field="expm1" fieldType="LRoot" ref="x" value="expm1" />
+        <putfield class="LRoot" field="expm1" fieldType="LRoot" ref="math" value="expm1" />
+        <new def="round" class="Ltensorflow/math/round" />
+        <putfield class="LRoot" field="round" fieldType="LRoot" ref="x" value="round" />
+        <putfield class="LRoot" field="round" fieldType="LRoot" ref="math" value="round" />
+        <new def="reciprocal" class="Ltensorflow/math/reciprocal" />
+        <putfield class="LRoot" field="reciprocal" fieldType="LRoot" ref="math" value="reciprocal" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:880]: > In this new binding for `reciprocal`, the previous `gen_math_ops.reciprocal` alias (formerly mapped via `pass_through`) was removed earlier in the file but is not restored here. That means code using `tf.gen_math_ops.reciprocal(...)` will no longer resolve to the modeled op. Consider adding a matching `<putfield ... ref="gen_math_ops" ...>` (and any other established aliases like `math_ops` if applicable) to avoid a behavior regression versus the earlier model.

Right &mdash; this was a regression in the alias surface. Restored the `gen_math_ops` binding in 53ad838a, mirroring the `tanh` pattern (`x` / `math` / `gen_math_ops` triple): https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L801-L803.

Holding off on `math_ops` until a fixture exercises it &mdash; the existing `gen_math_ops` alias is the one with documented runtime use; `math_ops` shows up in some compat shims but isn't a public TF surface. Easy to add later if needed.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 5]
```diff
@@ -784,6 +777,60 @@
...
+        <new def="softplus" class="Ltensorflow/math/softplus" />
+        <putfield class="LRoot" field="softplus" fieldType="LRoot" ref="math" value="softplus" />
+        <putfield class="LRoot" field="softplus" fieldType="LRoot" ref="nn" value="softplus" />
+        <new def="softsign" class="Ltensorflow/math/softsign" />
+        <putfield class="LRoot" field="softsign" fieldType="LRoot" ref="math" value="softsign" />
+        <putfield class="LRoot" field="softsign" fieldType="LRoot" ref="nn" value="softsign" />
+        <new def="square" class="Ltensorflow/math/square" />
+        <putfield class="LRoot" field="square" fieldType="LRoot" ref="x" value="square" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:889]: > `square` is rebound to the dedicated `Ltensorflow/math/square` class, but the prior `gen_math_ops.square` alias (previously provided via `pass_through`) is no longer present. If code calls `tf.gen_math_ops.square`, this will now fail to resolve or fall back to a less precise path. Consider also wiring `square` into `gen_math_ops` (and any other alias modules used for similar unary ops) for consistency with nearby ops like `tanh`/`cos` which include `gen_math_ops` bindings.

Same fix &mdash; restored the `gen_math_ops` binding in 53ad838a: https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L811-L813.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 6]
```diff
@@ -784,6 +777,60 @@
...
+        <putfield class="LRoot" field="square" fieldType="LRoot" ref="math" value="square" />
+        <new def="erf" class="Ltensorflow/math/erf" />
+        <putfield class="LRoot" field="erf" fieldType="LRoot" ref="math" value="erf" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:893]: > `erf` is now only bound under `math`, but the earlier `gen_math_ops.erf` alias was removed and not restored. This can regress modeling for code that references `tf.gen_math_ops.erf(...)`. Consider adding the corresponding `gen_math_ops` binding (and/or other established aliases) alongside the `math` binding.

Same fix &mdash; restored the `gen_math_ops` binding in 53ad838a: https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L815-L816.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 7]
```diff
@@ -784,6 +777,60 @@
...
+        <putfield class="LRoot" field="erfc" fieldType="LRoot" ref="math" value="erfc" />
+        <!-- Element-wise binary ops (broadcast shape, unified dtype) — routed through ElementWiseOperation. -->
+        <new def="atan2" class="Ltensorflow/math/atan2" />
+        <putfield class="LRoot" field="atan2" fieldType="LRoot" ref="math" value="atan2" />
+        <new def="maximum" class="Ltensorflow/math/maximum" />
+        <putfield class="LRoot" field="maximum" fieldType="LRoot" ref="x" value="maximum" />
+        <putfield class="LRoot" field="maximum" fieldType="LRoot" ref="math" value="maximum" />
+        <new def="minimum" class="Ltensorflow/math/minimum" />
+        <putfield class="LRoot" field="minimum" fieldType="LRoot" ref="x" value="minimum" />
+        <putfield class="LRoot" field="minimum" fieldType="LRoot" ref="math" value="minimum" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:906]: > `maximum`/`minimum` are introduced as dedicated classes and dispatched to `ElementWiseOperation`, but here they are only wired into `tf` (`ref="x"`) and `tf.math`. The existing `tf.gen_math_ops.maximum/minimum` aliases still point at the generic `pass_through` function earlier in the model, so calls through that alias will not get broadcast-shape/unified-dtype behavior. Consider also binding these ops into `gen_math_ops` (and any other alias modules used for similar elementwise ops) to keep behavior consistent across TensorFlow's multiple entry points.

Good catch &mdash; this was the most impactful of the four since the stale `pass_through` putfields *were* still in scope, so `tf.gen_math_ops.maximum(x, y)` was getting passthrough-on-the-second-operand behavior while `tf.maximum(x, y)` was getting the new `ElementWiseOperation` (broadcast shape + unified dtype) routing &mdash; a behavior asymmetry across alias surfaces. Fixed both in 53ad838a:

- Added the new `gen_math_ops` bindings to the dedicated classes: https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L823-L829.
- Deleted the stale `pass_through`-based `maximum`/`minimum` `gen_math_ops` putfields earlier in the file so the new bindings are the only ones in scope: https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L321-L328.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/types/TensorFlowTypes.java [hunk 8]
```diff
@@ -968,6 +968,190 @@ public boolean canConvertTo(DType other) {
 
   private static final String EXTRACT_PATCHES_SIGNATURE = "tf.image.extract_patches()";
 
+  // Tier-A pure-passthrough math ops (continued, wala/ML#422). Each is shape and dtype passthrough
+  // on its primary tensor argument (named `x` for most ops; `features` for `softplus` / `softsign`;
+  // and the second tensor operand of the binary ops `atan2` / `maximum` / `minimum`).
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/types/TensorFlowTypes.java]: > The comment here says these ops are shape/dtype passthrough on a "primary" argument and specifically calls out using the *second* tensor operand for the binary ops. However, `maximum`/`minimum`/`atan2` are dispatched to `ElementWiseOperation`, which derives dtype from the `x` parameter at position 0 and computes broadcast shape from both operands (not passthrough, and not "second operand"). Updating this comment to reflect the actual generator behavior would prevent future confusion when adding more elementwise ops.

Right &mdash; the "second tensor operand" phrasing was leftover from an earlier draft of this section. Rewrote the comment in 53ad838a to call out that the binary ops are the exception and route through `ElementWiseOperation` (dtype from `x`-position-0, broadcast shape from both operands): https://github.com/ponder-lab/ML/blob/53ad838a41ad599f83f6606811329cecce91518a/com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/types/TensorFlowTypes.java#L971-L975.

File: com.ibm.wala.cast.python.test/data/tf2_test_atan2.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import tensorflow as tf
+
+
+def f(a):
+    pass
+
+
+x = tf.constant([0.0, 0.5, 1.0])
+y_in = tf.constant([1.0, 1.0, 1.0])
+y = tf.math.atan2(x, y_in)
```
Comment (khatchad) [inline com.ibm.wala.cast.python.test/data/tf2_test_atan2.py:11]: Do we have tests with keyword arguments and a mixture of positional and keyword arguments?

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 10]
```diff
@@ -784,6 +768,65 @@
         <!-- https://www.tensorflow.org/versions/r2.9/api_docs/python/tf/custom_gradient -->
         <new def="custom_gradient" class="Ltensorflow/class/custom_gradient" />
         <putfield class="LRoot" field="custom_gradient" fieldType="LRoot" ref="x" value="custom_gradient" />
+        <!-- Tier-A pure-passthrough math ops (continued, wala/ML#422). Each shape and dtype passthrough on `x`. -->
+        <new def="tan" class="Ltensorflow/math/tan" />
+        <putfield class="LRoot" field="tan" fieldType="LRoot" ref="math" value="tan" />
+        <new def="asin" class="Ltensorflow/math/asin" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:846]: > `tan` and `asin` are only bound under `tf.math` here, but similar math ops in this model are typically also re-exported at the top-level `tf.*` (via `ref="x"`) and often under `tf.gen_math_ops.*`. As written, `tf.tan(...)` / `tf.asin(...)` (and any `gen_math_ops` alias if expected) won't resolve to ...

Right &mdash; `tf.tan(...)` and `tf.asin(...)` are real public APIs and `tf.gen_math_ops.{tan,asin}` are documented aliases. Added the missing bindings in 3b0f2710, mirroring `tanh`'s `x` / `math` / `gen_math_ops` triple: https://github.com/ponder-lab/ML/blob/3b0f271029a8f0b241091fcfa90ef76ed0e0c84a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L774-L781.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 11]
```diff
@@ -784,6 +768,65 @@
...
+        <putfield class="LRoot" field="asin" fieldType="LRoot" ref="math" value="asin" />
+        <new def="atan" class="Ltensorflow/math/atan" />
+        <putfield class="LRoot" field="atan" fieldType="LRoot" ref="math" value="atan" />
+        <new def="sinh" class="Ltensorflow/math/sinh" />
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml:854]: > `atan` and `sinh` are only wired under `tf.math` in this import block, while other math ops in the same section (e.g., `cosh`, `asinh`, `log1p`, etc.) are also wired to top-level `tf.*` via `ref="x"`. If `tf.atan(...)` / `tf.sinh(...)` are used in fixtures or real code, they won't dispatch to the ne...

Same fix &mdash; added `ref="x"` and `ref="gen_math_ops"` for `atan` and `sinh` in 3b0f2710: https://github.com/ponder-lab/ML/blob/3b0f271029a8f0b241091fcfa90ef76ed0e0c84a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L782-L789.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 12]
```diff
@@ -784,6 +768,65 @@
         <!-- https://www.tensorflow.org/versions/r2.9/api_docs/python/tf/custom_gradient -->
         <new def="custom_gradient" class="Ltensorflow/class/custom_gradient" />
         <putfield class="LRoot" field="custom_gradient" fieldType="LRoot" ref="x" value="custom_gradient" />
+        <!-- Tier-A pure-passthrough math ops (continued, wala/ML#422). Each shape and dtype passthrough on `x`. -->
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml]: > This comment says all ops in this block are shape/dtype passthrough on `x`, but the block includes ops whose primary tensor parameter is not named `x` (e.g., `softplus`/`softsign` use `features`) and also includes elementwise binary ops later. Updating the wording would prevent confusion when matchi...

Reworded the block header in 3b0f2710 to call out both exceptions: `softplus` / `softsign` use `features`, and the binary ops `atan2` / `maximum` / `minimum` route through `ElementWiseOperation` (not passthrough). https://github.com/ponder-lab/ML/blob/3b0f271029a8f0b241091fcfa90ef76ed0e0c84a/com.ibm.wala.cast.python.ml/data/tensorflow.xml#L771-L772

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 13]
```diff
@@ -4691,6 +4691,176 @@ public void testExtractPatches()
...
+
+  /**
+   * Keyword-argument variant of {@link #testMinimum}: {@code tf.math.minimum(x=..., y=...)}.
+   * Exercises the kw-arg-resolution path on the {@code ElementWiseOperation} dispatch.
+   */
+  @Test
+  public void testMinimumKw()
+      throws ClassHierarchyException, IllegalArgumentException, CancelException, IOException {
+    test("tf2_test_minimum_kw.py", "f", 1, 1, Map.of(2, Set.of(TENSOR_3_FLOAT32)));
+  }
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:5130]: > The PR description/test plan mentions "20 new tests" and "20 Python fixtures", but this change set adds additional keyword-argument variants as separate fixtures/tests (e.g., `testAtan2Kw`/`testMaximumKw`/`testMinimumKw` and the corresponding `tf2_test_*_kw.py` files). Consider updating the PR descr...

Updating the PR description to reflect the 3 kwargs-variant fixtures + tests added in `23eeaef5` (atan2 / maximum / minimum) and the `ceil_pair` 2-arg-sink canary added in `dfc5ccd8` &mdash; both arrived after the original "20 new tests" body was written.

File: com.ibm.wala.cast.python.ml/data/tensorflow.xml [hunk 14]
```diff
@@ -343,27 +343,11 @@
...
-        <putfield class="LRoot" field="square" fieldType="LRoot" ref="gen_math_ops" value="pass_through" />
-        <!-- https://www.tensorflow.org/versions/r2.9/api_docs/python/tf/math/maximum -->
-        <putfield class="LRoot" field="maximum" fieldType="LRoot" ref="x" value="pass_through" />
-        <putfield class="LRoot" field="maximum" fieldType="LRoot" ref="math" value="pass_through" />
-        <putfield class="LRoot" field="maximum" fieldType="LRoot" ref="gen_math_ops" value="pass_through" />
-        <!-- https://www.tensorflow.org/versions/r2.9/api_docs/python/tf/math/minimum -->
-        <putfield class="LRoot" field="minimum" fieldType="LRoot" ref="x" value="pass_through" />
-        <putfield class="LRoot" field="minimum" fieldType="LRoot" ref="math" value="pass_through" />
-        <putfield class="LRoot" field="minimum" fieldType="LRoot" ref="gen_math_ops" value="pass_through" />
+        <!-- `tf.{reciprocal,erf,square}` and `tf.{math,gen_math_ops}.{reciprocal,erf,square}` now route through dedicated `<class>` blocks defined later in this file (search for `class name="reciprocal"`, etc.). The new bindings include the `gen_math_ops` alias as well. -->
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/data/tensorflow.xml]: Fixed in 7c51f06b. Updated the comment to reflect the actual alias surface: `tf.math.{reciprocal,erf}` and `tf.gen_math_ops.{reciprocal,erf}` (no top-level `tf.reciprocal` / `tf.erf` because those aren't real TF 2 names), while keeping `tf.square` in the top-level claim (it IS a real TF 2 top-level and the matching `<putfield ref="x">` is present).
